walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,146 @@
 
 package com.google.sps;
 
+import java.util.Collections;
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // This variable will hold all of the invalid meeting times.
+    // This list is sorted by start times, and has no overlapping times.
+    List<TimeRange> invalidTimeRanges = new ArrayList<>();
+    
+    // This is the list of the attendees which attend the meeting.
+    Collection<String> attendees = request.getAttendees();
+
+    // The duration of the meeting request.
+    int meetingDuration = (int) request.getDuration();
+
+    // Populate invalidTimeRanges variable, and sort by increasing start time.
+    populateInvalidTimeRanges(invalidTimeRanges, events, attendees);
+    Collections.sort(invalidTimeRanges, TimeRange.ORDER_BY_START);
+
+    // Merge conflicting TimeRanges.
+    mergeConflictingTimeRanges(invalidTimeRanges);
+
+    // Get the inverse / valid TimeRanges of the invalidTimeRanges.
+    List<TimeRange> validTimeRanges = getValidTimeRanges(invalidTimeRanges, 
+      meetingDuration);
+
+    return validTimeRanges;
+
+  }
+
+	/**
+   * Returns true if the duration (presumably of a TimeRange) is longer than
+   * the meeting duration request.
+   */
+  private boolean fitsMeetingDuration(int duration, int meetingDuration) {
+    return duration >= meetingDuration; 
+  }
+
+  /**
+   * Get the valid (inverse) TimeRanges from invalidTimeRanges.
+   * A valid TimeRange is not added if it is shorter than the meeting request 
+   * length.
+   * 
+   * Example:
+   * Meeting Request Duration: 30 minutes.
+   * invalidTimeRanges: [(100, 200), (500, 1400), (1700, 1900)]
+   * validTimeRanges:   [(0, 100), (200, 500), (1400, 1700), (1900, 2400)]
+   */
+  private List<TimeRange> getValidTimeRanges(List<TimeRange> invalidTimeRanges, 
+    int meetingDuration) {
+    List<TimeRange> validTimeRanges = new ArrayList<>();
+
+    // Get the inverse TimeRanges of the invalidTimeRanges.
+    int index = 0;
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.START_OF_DAY;
+    while (index < invalidTimeRanges.size()) {
+      end = invalidTimeRanges.get(index).start();
+
+      // Add new valid TimeRange to list.
+      if (start != end && fitsMeetingDuration(end - start, meetingDuration)) {
+        validTimeRanges.add(TimeRange.fromStartEnd(start, end, false));
+      }
```

Comment:
If you're going to add a helper method here, you could put a lot more in. This whole block is used again below for the TimeRange.END_OF_DAY, so why not make a maybeAddTimeRange method?

private void maybeAddTimeRange(List< TimeRange> validTimeRanges, int start, int end, int meeting duration, boolean inclusive) {
    if (start != end && (end - start) >= meetingDuration)) {
        validTimeRanges.add(TimeRange.fromStartEnd(start, end, inclusive));
      }
}